File: snippets/security/sanitize-middleware.py
"""
SNIPPET: Global Sanitization Middleware
CATEGORY: Security
LANGUAGE: Python (FastAPI)
STATUS: Ready

DESCRIPTION:
    Global middleware that automatically sanitizes ALL incoming request bodies
    and ALL outgoing JSON responses. Acts as a safety net — even if per-route
    validation is forgotten, this catches dangerous input/output.

    Also includes validation helpers: is_positive_int, is_allowed_value,
    validate_required_fields, enforce_max_lengths.

DEPENDENCIES:
    pip install nh3

USAGE:
    from sanitize_middleware import SanitizeMiddleware, is_positive_int

    app = FastAPI()
    app.add_middleware(SanitizeMiddleware)

    @app.get("/items/{item_id}")
    async def get_item(item_id: int):
        if not is_positive_int(item_id):
            raise HTTPException(status_code=400, detail="Invalid ID")
        ...

IMPORTANT:
    - Add SanitizeMiddleware BEFORE defining routes
    - Pydantic models handle most validation, but this middleware is a safety net
    - These are a SAFETY NET, not a replacement for Pydantic validation
"""

import json
from typing import Any
import nh3
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import Response, JSONResponse
from fastapi import FastAPI


# ---------------------------------------------------------------------------
# Sanitization functions
# ---------------------------------------------------------------------------

def strip_html(value: str) -> str:
    """
    Strip ALL HTML tags from a string.
    Uses nh3 library (Rust-backed, replaces deprecated bleach) for robust
    protection against XSS bypasses like unclosed tags and nested encoding.
    """
    if not isinstance(value, str):
        return value
    return nh3.clean(value, tags=set())


def deep_sanitize(obj: Any) -> Any:
    """Recursively sanitize all string values in an object/list."""
    if isinstance(obj, str):
        return strip_html(obj).strip()
    if isinstance(obj, list):
        return [deep_sanitize(item) for item in obj]
    if isinstance(obj, dict):
        return {key: deep_sanitize(value) for key, value in obj.items()}
    return obj


# ---------------------------------------------------------------------------
# Global Sanitization Middleware
# ---------------------------------------------------------------------------

class SanitizeMiddleware(BaseHTTPMiddleware):
    """
    Middleware that sanitizes both request bodies and response bodies.
    - Strips HTML tags from all string values in JSON request bodies
    - Strips HTML tags from all string values in JSON responses
    """

    async def dispatch(self, request: Request, call_next) -> Response:
        # --- Sanitize request body ---
        if request.method in ("POST", "PUT", "PATCH"):
            content_type = request.headers.get("content-type", "")
            if "application/json" in content_type:
                try:
                    body = await request.body()
                    if body:
                        parsed = json.loads(body)
                        sanitized = deep_sanitize(parsed)
                        # Replace the request body with sanitized version
                        request._body = json.dumps(sanitized).encode()
                except (json.JSONDecodeError, UnicodeDecodeError):
                    pass  # Let FastAPI handle malformed JSON

        # --- Call the route handler ---
        response = await call_next(request)

        # --- Sanitize response body ---
        if response.headers.get("content-type", "").startswith("application/json"):
            body = b""
            async for chunk in response.body_iterator:
                if isinstance(chunk, str):
                    chunk = chunk.encode()
                body += chunk
            try:
                parsed = json.loads(body)
                sanitized = deep_sanitize(parsed)
                return JSONResponse(
                    content=sanitized,
                    status_code=response.status_code,
                    headers=dict(response.headers),
                )
            except (json.JSONDecodeError, UnicodeDecodeError):
                pass  # Return original response if not valid JSON

        return response


# ---------------------------------------------------------------------------
# Global Error Handler
# ---------------------------------------------------------------------------

def add_global_error_handler(app: FastAPI) -> None:
    """
    Add a global exception handler that never exposes stack traces.
    Returns generic message for 500s.

    Usage:
        app = FastAPI()
        add_global_error_handler(app)
    """

    @app.exception_handler(Exception)
    async def global_exception_handler(request: Request, exc: Exception):
        import traceback
        traceback.print_exc()  # Log full error server-side
        return JSONResponse(
            status_code=500,
            content={
                "error": "INTERNAL_ERROR",
                "message": "An unexpected error occurred",
            },
        )


# ---------------------------------------------------------------------------
# Per-route validation helpers
# ---------------------------------------------------------------------------

def is_positive_int(value) -> bool:
    """Check if a value is a positive integer (for :id params)."""
    try:
        num = int(value)
        return num > 0
    except (TypeError, ValueError):
        return False


def is_allowed_value(value: str, allowed: list[str]) -> bool:
    """Check if a value is in an allowed list (for query params / enums)."""
    return value in allowed


def validate_required_fields(body: dict, fields: list[str]) -> list[str]:
    """
    Validate that required fields are present.
    Returns list of missing field names, or empty list if all present.
    """
    return [f for f in fields if not body.get(f)]


def enforce_max_lengths(body: dict, max_lengths: dict[str, int]) -> list[dict]:
    """
    Enforce max lengths on string fields.
    Returns list of violations: [{"field": name, "max": N, "actual": N}]
    """
    violations = []
    for field, max_len in max_lengths.items():
        value = body.get(field)
        if isinstance(value, str) and len(value) > max_len:
            violations.append({"field": field, "max": max_len, "actual": len(value)})
    return violations


# ---------------------------------------------------------------------------
# Setup helper
# ---------------------------------------------------------------------------

def apply_security_middleware(app: FastAPI) -> None:
    """
    Apply all global security middleware to a FastAPI app.
    Call this ONCE during app setup, BEFORE defining routes.

    Usage:
        from sanitize_middleware import apply_security_middleware

        app = FastAPI()
        apply_security_middleware(app)
    """
    app.add_middleware(SanitizeMiddleware)
    add_global_error_handler(app)

File: snippets/security/sanitize-middleware.js
/**
 * SNIPPET: Global Sanitization Middleware
 * CATEGORY: Security
 * LANGUAGE: JavaScript (Express)
 * STATUS: ✅ Ready
 *
 * DESCRIPTION:
 *   Global middleware that automatically sanitizes ALL incoming request bodies
 *   and ALL outgoing JSON responses. Acts as a safety net — even if per-route
 *   validation is forgotten, this catches dangerous input/output.
 *
 *   Also includes per-route helpers: isPositiveInt, isAllowedValue,
 *   validateRequiredFields, enforceMaxLengths.
 *
 * DEPENDENCIES:
 *   npm install sanitize-html
 *
 * USAGE:
 *   const { sanitizeBody, sanitizeResponse } = require('./middleware/sanitize');
 *
 *   // Apply BEFORE routes
 *   app.use(express.json({ limit: '100kb' }));
 *   app.use(sanitizeBody);
 *   app.use(sanitizeResponse);
 *
 *   // Per-route helpers
 *   app.get('/items/:id', (req, res) => {
 *     if (!isPositiveInt(req.params.id)) return res.status(400).json({ error: 'Invalid ID' });
 *     ...
 *   });
 *
 * IMPORTANT:
 *   - Apply sanitizeBody AFTER express.json() but BEFORE routes
 *   - Apply sanitizeResponse BEFORE routes (it wraps res.json)
 *   - Set express.json({ limit: '100kb' }) to prevent large payload attacks
 *   - These are a SAFETY NET, not a replacement for per-route validation
 */

// ============================================
// GLOBAL SANITIZATION MIDDLEWARE
// ============================================

const sanitizeHtml = require('sanitize-html');

/**
 * Strip ALL HTML tags from a string value.
 * Uses sanitize-html library (not regex) for robust protection against
 * XSS bypasses like unclosed tags, nested encoding, and attribute injection.
 * @param {string} str
 * @returns {string}
 */
function stripHtml(str) {
  if (typeof str !== 'string') return str;
  return sanitizeHtml(str, { allowedTags: [], allowedAttributes: {} });
}

/**
 * Recursively sanitize all string values in an object/array.
 * Strips HTML tags and trims whitespace.
 * @param {*} obj
 * @returns {*}
 */
function deepSanitize(obj) {
  if (typeof obj === 'string') {
    return stripHtml(obj).trim();
  }
  if (Array.isArray(obj)) {
    return obj.map(deepSanitize);
  }
  if (obj && typeof obj === 'object') {
    const sanitized = {};
    for (const [key, value] of Object.entries(obj)) {
      sanitized[key] = deepSanitize(value);
    }
    return sanitized;
  }
  return obj;
}

/**
 * Middleware: sanitize all string values in req.body.
 * Apply AFTER express.json(), BEFORE routes.
 */
function sanitizeBody(req, res, next) {
  if (req.body && typeof req.body === 'object') {
    req.body = deepSanitize(req.body);
  }
  next();
}

/**
 * Middleware: sanitize all string values in JSON responses.
 * Wraps res.json() to strip HTML from outgoing data.
 */
function sanitizeResponse(req, res, next) {
  const originalJson = res.json.bind(res);
  res.json = (data) => {
    if (data && typeof data === 'object') {
      data = deepSanitize(data);
    }
    return originalJson(data);
  };
  next();
}

// ============================================
// GLOBAL ERROR HANDLER
// ============================================

/**
 * Global error handler — add AFTER all routes.
 * Returns generic message for 500s, never exposes stack traces.
 */
function globalErrorHandler(err, req, res, _next) {
  console.error('[Error]', err);

  const statusCode = err.statusCode || 500;
  res.status(statusCode).json({
    error: err.code || 'INTERNAL_ERROR',
    message: statusCode === 500
      ? 'An unexpected error occurred'
      : err.message,
  });
}

// ============================================
// PER-ROUTE VALIDATION HELPERS
// ============================================

/**
 * Check if a value is a positive integer (for :id params).
 * @param {string|number} value
 * @returns {boolean}
 */
function isPositiveInt(value) {
  const num = Number(value);
  return Number.isInteger(num) && num > 0;
}

/**
 * Check if a value is in an allowed list (for query params / enums).
 * @param {string} value
 * @param {string[]} allowed
 * @returns {boolean}
 */
function isAllowedValue(value, allowed) {
  return allowed.includes(value);
}

/**
 * Validate that required fields are present in req.body.
 * Returns array of missing field names, or empty array if all present.
 * @param {object} body - req.body
 * @param {string[]} fields - required field names
 * @returns {string[]} missing fields
 */
function validateRequiredFields(body, fields) {
  return fields.filter(f => body[f] === undefined || body[f] === null || body[f] === '');
}

/**
 * Enforce max lengths on string fields. Returns array of violations.
 * @param {object} body - req.body
 * @param {object} maxLengths - { fieldName: maxLength }
 * @returns {Array<{field: string, max: number, actual: number}>}
 */
function enforceMaxLengths(body, maxLengths) {
  const violations = [];
  for (const [field, max] of Object.entries(maxLengths)) {
    if (typeof body[field] === 'string' && body[field].length > max) {
      violations.push({ field, max, actual: body[field].length });
    }
  }
  return violations;
}

// ============================================
// SETUP HELPER
// ============================================

/**
 * Apply all global security middleware to an Express app.
 * Call this ONCE during app setup, BEFORE defining routes.
 *
 * @param {import('express').Express} app
 *
 * @example
 *   const express = require('express');
 *   const { applySecurityMiddleware } = require('./middleware/sanitize');
 *
 *   const app = express();
 *   applySecurityMiddleware(app);
 *
 *   // Define routes AFTER this
 *   app.get('/health', (req, res) => res.json({ status: 'ok' }));
 *
 *   // Add error handler AFTER routes
 *   app.use(globalErrorHandler);
 */
function applySecurityMiddleware(app) {
  // Limit payload size to prevent large payload attacks
  // Default: 1mb. Adjust based on your use case (file uploads use multer, not this)
  app.use(require('express').json({ limit: '1mb' }));
  app.use(require('express').urlencoded({ extended: false, limit: '1mb' }));

  // Sanitize all incoming request bodies
  app.use(sanitizeBody);

  // Sanitize all outgoing JSON responses
  app.use(sanitizeResponse);
}

module.exports = {
  // Global middleware
  sanitizeBody,
  sanitizeResponse,
  globalErrorHandler,
  applySecurityMiddleware,

  // Sanitization functions
  stripHtml,
  deepSanitize,

  // Per-route helpers
  isPositiveInt,
  isAllowedValue,
  validateRequiredFields,
  enforceMaxLengths,
};